morpheus/_lib/src/utilities/rest_server.cpp

```diff
@@ -0,0 +1,436 @@
+                ->run();
+        }
+
+        do_accept();
+    }
+
+    std::shared_ptr<boost::asio::io_context> m_io_context;
+    tcp::endpoint m_tcp_endpoint;
+    tcp::acceptor m_acceptor;
+
+    std::shared_ptr<morpheus::payload_parse_fn_t> m_payload_parse_fn;
+    const std::string& m_url_endpoint;
+    http::verb m_method;
+    std::size_t m_max_payload_size;
+    std::chrono::seconds m_request_timeout;
+};
+
+}  // namespace
+
+namespace morpheus {
+
+RestServer::RestServer(payload_parse_fn_t payload_parse_fn,
+                       std::string bind_address,
+                       unsigned short port,
+                       std::string endpoint,
+                       std::string method,
+                       unsigned short num_threads,
+                       std::size_t max_payload_size,
+                       std::chrono::seconds request_timeout) :
+  m_payload_parse_fn(std::make_shared<payload_parse_fn_t>(std::move(payload_parse_fn))),
+  m_bind_address(std::move(bind_address)),
+  m_port(port),
+  m_endpoint(std::move(endpoint)),
+  m_method(http::string_to_verb(method)),
+  m_num_threads(num_threads),
+  m_request_timeout(request_timeout),
+  m_max_payload_size(max_payload_size),
+  m_io_context{nullptr},
+  m_is_running{false}
+{
+    if (m_method == http::verb::unknown)
+    {
+        throw std::runtime_error("Invalid method: " + method);
+    }
+
+    if (m_num_threads == 0)
+    {
+        throw std::runtime_error("num_threads must be greater than 0");
+    }
+}
+
+void RestServer::start_listener()
+{
+    DCHECK(m_io_context == nullptr) << "start_listener expects m_io_context to be null";
+
+    // This function will block until the io context is shutdown, and should be called from the first worker thread
+    DCHECK(m_listener_threads.size() == 1 && m_listener_threads[0].get_id() == std::this_thread::get_id())
+        << "start_listener must be called from the first thread in m_listener_threads";
+
+    m_io_context = std::make_shared<net::io_context>(m_num_threads);
+    auto ioc     = m_io_context;  // ensure each thread gets its own copy including this one
+
+    std::make_shared<Listener>(
+        ioc, m_payload_parse_fn, m_bind_address, m_port, m_endpoint, m_method, m_max_payload_size, m_request_timeout)
+        ->run();
+
+    for (auto i = 1; i < m_num_threads; ++i)
+    {
+        m_listener_threads.emplace_back([ioc]() { ioc->run(); });
+    }
+
+    ioc->run();
+}
+
+void RestServer::start()
+{
+    CHECK(!m_is_running) << "RestServer is already running";
+
+    try
+    {
+        DLOG(INFO) << "Starting RestServer on " << m_bind_address << ":" << m_port << " with " << m_num_threads
+                   << " threads";
+        m_is_running = true;
+        m_listener_threads.reserve(m_num_threads);
+        m_listener_threads.emplace_back(std::thread(&RestServer::start_listener, this));
```

Comment:
In general, we want to use MRC to manage all threads/fibers so they can be managed by the runtime. But this should be ok. Going to look and see if there is any way to better link the boost thread creation with MRC.

Main thing I am worried about here is MRC starts every thread with some initializer code to allow for debugging python. Without it, breakpoints and things wont work. Since this does a callback into python, you wouldnt hit any breakpoints.